src/runner_metrics.py

```diff
@@ -0,0 +1,222 @@
+
+FILE_SIZE_BYTES_LIMIT = 1024
+PRE_JOB_METRICS_FILE_NAME = "pre-job-metrics.json"
+POST_JOB_METRICS_FILE_NAME = "post-job-metrics.json"
+RUNNER_INSTALLED_TS_FILE_NAME = "runner-installed.timestamp"
+
+
+class PreJobMetrics(BaseModel):
+    """Metrics for the pre-job phase of a runner.
+
+    Args:
+        timestamp: The UNIX time stamp of the time at which the event was originally issued.
+        workflow: The workflow name.
+        workflow_run_id: The workflow run id.
+        repository: The repository name.
+        event: The github event.
+    """
+
+    timestamp: NonNegativeFloat
+    workflow: str
+    workflow_run_id: str
+    repository: str
+    event: str
+
+
+class PostJobMetrics(BaseModel):
+    """Metrics for the post-job phase of a runner.
+
+    Args:
+        timestamp: The UNIX time stamp of the time at which the event was originally issued.
+        status: The status of the job.
+    """
+
+    timestamp: NonNegativeFloat
+    status: str
+
+
+class RunnerMetrics(BaseModel):
+    """Metrics for a runner.
+
+    Args:
+        installed_timestamp: The UNIX time stamp of the time at which the runner was installed.
+        pre_job: The metrics for the pre-job phase.
+        post_job: The metrics for the post-job phase.
+    """
+
+    installed_timestamp: NonNegativeFloat
+    pre_job: PreJobMetrics
+    post_job: Optional[PostJobMetrics]
+
+
+def _inspect_file_sizes(fs: shared_fs.SharedFilesystem) -> tuple[Path, ...]:
+    """Inspect the file sizes of the shared filesystem.
+
+    Args:
+        fs: The shared filesystem for a specific runner.
+
+    Returns:
+        A tuple of files whose size is larger than the limit.
+    """
+    files: list[Path] = [
+        fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(POST_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME),
+    ]
+
+    return tuple(
+        filter(lambda file: file.exists() and file.stat().st_size > FILE_SIZE_BYTES_LIMIT, files)
+    )
+
+
+def _extract_metrics_from_fs(fs: shared_fs.SharedFilesystem) -> Optional[RunnerMetrics]:
+    """Extract metrics from a shared filesystem.
+
+    Args:
+        fs: The shared filesystem for a specific runner.
+
+    Returns:
+        The extracted metrics if at least the pre-job metrics are present.
+
+    Raises:
+        CorruptMetricDataError: Raised if one of the files is not valid or too large.
+    """
+    if too_large_files := _inspect_file_sizes(fs):
+        raise CorruptMetricDataError(
+            f"File size of {too_large_files} is too large. "
+            f"The limit is {FILE_SIZE_BYTES_LIMIT} bytes."
+        )
+
+    try:
+        installed_timestamp = fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME).read_text()
+    except FileNotFoundError:
+        logger.exception("installed_timestamp not found for runner %s", fs.runner_name)
+        return None
+
+    logger.debug("Runner %s installed at %s", fs.runner_name, installed_timestamp)
```

Comment:
I would move the log inside the try block similarly to other metrics log afterwards